src/compiler/transformers/automatic-key-insertion/index.ts

```diff
@@ -0,0 +1,158 @@
+   *
+   * @param node a typescript syntax tree node
+   */
+  function findRenderMethodVisitor(node: ts.Node): ts.VisitResult<ts.Node> {
+    if (ts.isMethodDeclaration(node) && node.name.getText() === 'render') {
+      return ts.visitEachChild(node, jsxElementVisitor, transformCtx);
+    } else {
+      return ts.visitEachChild(node, findRenderMethodVisitor, transformCtx);
+    }
+  }
+
+  /**
+   * Our inner-most visitor function. This will edit any JSX nodes that it
+   * finds, adding a `key` attribute to them via {@link addKeyAttr}.
+   *
+   * @param node a typescript syntax tree node
+   */
+  function jsxElementVisitor(node: ts.Node): ts.VisitResult<ts.Node> {
+    if (isJSXElWithAttrs(node)) {
+      return addKeyAttr(node);
+    } else {
+      return ts.visitEachChild(node, jsxElementVisitor, transformCtx);
+    }
+  }
+
+  return (tsSourceFile) => {
+    return ts.visitEachChild(tsSourceFile, findClassDeclVisitor, transformCtx);
+  };
+};
+
+/**
+ * An incrementing-number generator, just as a little extra 'uniqueness'
+ * insurance for {@link deriveJSXKey}
+ */
+const incrementer = (function* () {
+  let val = 0;
+  while (true) {
+    yield val++;
+  }
+})();
+
+/**
+ * Generate a unique key for a given JSX element. The key is creating by
+ * concatenating and then hashing (w/ SHA1) the following:
+ *
+ * - an incrementing value
+ * - the element's tag name
+ * - the start position for the element's token in the original source file
+ * - the end position for the element's token in the original source file
+ *
+ * It is hoped this provides enough uniqueness that a collision won't occur.
+ *
+ * @param jsxElement a typescript JSX syntax tree node which needs a key
+ * @returns a computed unique key for that element
+ */
+function deriveJSXKey(jsxElement: ts.JsxOpeningElement | ts.JsxSelfClosingElement): string {
+  const hash = createHash('sha1')
+    .update(`${incrementer.next().value}__${jsxElement.tagName}__${jsxElement.pos}_${jsxElement.end}`)
+    .digest('hex')
+    .toLowerCase();
+  return hash;
+}
+
+/**
+ * description
+ *
+ * @param node a typescript syntax tree node
+ */
+function isJSXElWithAttrs(node: ts.Node): node is ts.JsxOpeningElement | ts.JsxSelfClosingElement {
+  return ts.isJsxOpeningElement(node) || ts.isJsxSelfClosingElement(node);
+}
+
+/**
+ * Given a JSX syntax tree node update it to include a unique key attribute.
+ *
+ * @param jsxElement a typescript JSX syntax tree node
+ */
+function addKeyAttr(
+  jsxElement: ts.JsxOpeningElement | ts.JsxSelfClosingElement,
+): ts.JsxOpeningElement | ts.JsxSelfClosingElement {
+  const attrs = ts.factory.createJsxAttributes([
+    ts.factory.createJsxAttribute(
+      ts.factory.createIdentifier('key'),
+      ts.factory.createStringLiteral(deriveJSXKey(jsxElement)),
+    ),
+    ...jsxElement.attributes.properties,
+  ]);
+
+  if (jsxElement.tagName.getText() === 'rect' || jsxElement.tagName.getText() === 'svg') {
+    // TODO diagnose the issue with svg tags
```

Comment:
svg attrs are _sometimes_ being dropped without this check, I'm not sure why but I'm going to leave this for now so I remember to investigate fully and then fix that issue